worker/upload.go

```diff
@@ -353,33 +354,83 @@ func (w *worker) tryUploadPackedSlabs(ctx context.Context, rs api.RedundancySett
 	// if we want to block, try and upload one packed slab synchronously, we use
 	// a slightly higher upload priority to avoid reaching the context deadline
 	if block {
-		_, err = w.uploadPackedSlabs(ctx, defaultPackedSlabsLockDuration, rs, contractSet, defaultPackedSlabsLimit, lockingPriorityBlockedUpload)
+		_, err = w.uploadPackedSlabs(ctx, defaultPackedSlabsLockDuration, rs, contractSet, lockingPriorityBlockedUpload)
 	}
 
 	// make sure there's a goroutine uploading the remainder of the packed slabs
 	go w.threadedUploadPackedSlabs(rs, contractSet, lockingPriorityBackgroundUpload)
 	return
 }
 
-func (w *worker) uploadPackedSlabs(ctx context.Context, lockingDuration time.Duration, rs api.RedundancySettings, contractSet string, limit, lockPriority int) (uploaded int, err error) {
-	// fetch packed slabs
-	packedSlabs, err := w.bus.PackedSlabsForUpload(ctx, lockingDuration, uint8(rs.MinShards), uint8(rs.TotalShards), contractSet, limit)
-	if err != nil {
-		return 0, fmt.Errorf("couldn't fetch packed slabs from bus: %v", err)
-	}
-
+func (w *worker) uploadPackedSlabs(ctx context.Context, lockingDuration time.Duration, rs api.RedundancySettings, contractSet string, lockPriority int) (uploaded int, err error) {
 	// upload packed slabs
-	for _, ps := range packedSlabs {
-		err = w.uploadPackedSlab(ctx, ps, rs, contractSet, lockPriority)
+	var mu sync.Mutex
+	errChan := make(chan error, 1)
+
+	var wg sync.WaitGroup
+	redundantSize := uint64(rs.TotalShards) * rhpv2.SectorSize
+
+LOOP:
+	for {
+		// block until either an error occurs or the next slab is ready to be
+		// uploaded
+		var mem *acquiredMemory
+		select {
+		case <-ctx.Done():
+			break LOOP
+		case err = <-errChan:
+			break LOOP
+		case mem = <-w.uploadManager.mm.AcquireMemory(ctx, redundantSize):
+		}
+		if mem == nil {
+			break // interrupted
+		}
+
+		// fetch packed slabs to upload
+		var packedSlabs []api.PackedSlab
+		packedSlabs, err = w.bus.PackedSlabsForUpload(ctx, lockingDuration, uint8(rs.MinShards), uint8(rs.TotalShards), contractSet, 1)
 		if err != nil {
-			return
+			err = fmt.Errorf("couldn't fetch packed slabs from bus: %v", err)
+			mem.Release()
+			break LOOP
+		} else if len(packedSlabs) == 0 {
+			mem.Release()
+			break LOOP // no more slabs
 		}
-		uploaded++
+		ps := packedSlabs[0]
+
+		// launch upload for slab
+		wg.Add(1)
+		go func(ps api.PackedSlab) {
+			defer mem.Release()
+			defer wg.Done()
+			err := w.uploadPackedSlab(ctx, ps, rs, contractSet, lockPriority, mem)
+			if err != nil {
+				select {
+				case errChan <- err:
+				default:
+				}
+			} else {
+				mu.Lock()
+				uploaded++
+				mu.Unlock()
+			}
+		}(ps)
+	}
+
+	// wait for all threads to finish
+	wg.Wait()
```

Comment:
Why is this wait necessary? I figured you were deferring the close of a channel but I don't see `errChan` get closed actually. If we know we are going to fail we might fail immediately?

Reply:
To make sure all uploads are done and we catch a potential error that might otherwise go unnoticed